File: src/dataStructures/priorityQueue.ts
/** How the comparison function needs to work */
export type PriorityCompare<T> = (a: T, b: T) => number;

/** A Priority Queue */
export default class PriorityQueue<T = number> {
  #length: number = 0;
  /**
   * @param data - initial data
   * @param compare - compare function
   */
  constructor(
    private data: T[] = [],
    private compare: PriorityCompare<T> = (a: T, b: T): number => (a < b ? -1 : a > b ? 1 : 0),
  ) {
    this.#length = data.length;
    if (this.#length > 0) {
      for (let i = (this.#length >> 1) - 1; i >= 0; i--) this.#down(i);
    }
  }

  /** @returns - the number of items */
  get length(): number {
    return this.#length;
  }

  /** @param item - the item to store */
  push(item: T): void {
    this.data.push(item);
    this.#up(this.#length++);
  }

  /**
   * Access the top item, remove it, and return it
   * @returns - the top item
   */
  pop(): T | undefined {
    if (this.#length === 0) return;
    const top = this.data.at(0);
    const bottom = this.data.pop();
    if (bottom === undefined) return;

    if (--this.#length > 0) {
      this.data[0] = bottom;
      this.#down(0);
    }

    return top;
  }

  /**
   * Peek at the top item without removing it
   * @returns - the top item
   */
  peek(): T | undefined {
    return this.data.at(0);
  }

  /**
   * @param pos - the position to reorder
   */
  #up(pos: number): void {
    const { data, compare } = this;
    const item = data[pos];

    while (pos > 0) {
      const parent = (pos - 1) >> 1;
      const current = data[parent];
      if (compare(item, current) >= 0) break;
      data[pos] = current;
      pos = parent;
    }

    data[pos] = item;
  }

  /** @param pos - the position to start reordering */
  #down(pos: number): void {
    const { data, compare } = this;
    const halfLength = this.#length >> 1;
    const item = data[pos];

    while (pos < halfLength) {
      let bestChild = (pos << 1) + 1; // initially it is the left child
      const right = bestChild + 1;

      if (right < this.#length && compare(data[right], data[bestChild]) < 0) {
        bestChild = right;
      }
      if (compare(data[bestChild], item) >= 0) break;

      data[pos] = data[bestChild];
      pos = bestChild;
    }

    data[pos] = item;
  }
}

File: rust/data_structures/priority_queue.rs
use alloc::vec::Vec;

/// Comparison function type for priority queue elements.
pub type PriorityCompare<T> = fn(&T, &T) -> core::cmp::Ordering;

/// # Priority Queue
///
/// ## Description
/// A priority queue is a data structure that stores elements in a specific order.
///
/// ## Usage
/// ```rust
/// use gistools::data_structures::PriorityQueue;
///
/// let mut queue = PriorityQueue::new(|a: &i32, b: &i32| a.cmp(b));
///
/// queue.push(3);
/// queue.push(1);
/// queue.push(2);
///
/// assert_eq!(queue.peek(), Some(&1));
/// assert_eq!(queue.pop(), Some(1));
/// assert_eq!(queue.len(), 2);
/// ```
pub struct PriorityQueue<T> {
    data: Vec<T>,
    compare: PriorityCompare<T>,
}

impl<T> PriorityQueue<T>
where
    T: Ord + Clone,
{
    /// Creates a new priority queue with an optional comparison function.
    pub fn new(compare: PriorityCompare<T>) -> Self {
        Self { data: Vec::new(), compare }
    }

    /// Returns the number of elements in the queue.
    pub fn len(&self) -> usize {
        self.data.len()
    }

    /// Returns true if the queue is empty.
    pub fn is_empty(&self) -> bool {
        self.data.is_empty()
    }

    /// Pushes an item into the queue.
    pub fn push(&mut self, item: T) {
        self.data.push(item);
        self.up(self.data.len() - 1);
    }

    /// Removes and returns the highest-priority item.
    pub fn pop(&mut self) -> Option<T> {
        if self.data.is_empty() {
            return None;
        }
        let top = self.data.swap_remove(0);
        if !self.data.is_empty() {
            self.down(0);
        }
        Some(top)
    }

    /// Peeks at the highest-priority item without removing it.
    pub fn peek(&self) -> Option<&T> {
        self.data.first()
    }

    fn up(&mut self, mut pos: usize) {
        let item = self.data[pos].clone();
        while pos > 0 {
            let parent = (pos - 1) / 2;
            if (self.compare)(&self.data[parent], &item) != core::cmp::Ordering::Greater {
                break;
            }
            self.data[pos] = self.data[parent].clone();
            pos = parent;
        }
        self.data[pos] = item;
    }

    fn down(&mut self, mut pos: usize) {
        let len = self.data.len();
        let item = self.data[pos].clone();
        let half_len = len / 2;
        while pos < half_len {
            let mut child = 2 * pos + 1;
            if child + 1 < len
                && (self.compare)(&self.data[child + 1], &self.data[child])
                    == core::cmp::Ordering::Less
            {
                child += 1;
            }
            if (self.compare)(&self.data[child], &item) != core::cmp::Ordering::Less {
                break;
            }
            self.data[pos] = self.data[child].clone();
            pos = child;
        }
        self.data[pos] = item;
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_priority_queue() {
        let mut queue = PriorityQueue::new(|a: &i32, b: &i32| a.cmp(b));
        assert!(queue.is_empty());
        assert_eq!(queue.peek(), None);
        assert_eq!(queue.pop(), None);
        queue.push(3);
        queue.push(1);
        queue.push(2);
        assert!(!queue.is_empty());
        assert_eq!(queue.peek(), Some(&1));
        assert_eq!(queue.pop(), Some(1));
        assert_eq!(queue.len(), 2);
    }

    #[test]
    fn test_priority_queue_large() {
        let mut queue = PriorityQueue::new(|a: &i32, b: &i32| a.cmp(b));

        // add 200 elements
        for i in 0..200 {
            queue.push(i);
        }

        assert_eq!(queue.pop(), Some(0));
    }

    #[test]
    fn test_priority_queue_large_inverse() {
        let mut queue = PriorityQueue::new(|a: &i32, b: &i32| a.cmp(b));

        // add 200 elements
        for i in 0..200 {
            queue.push(200 - i);
        }

        assert_eq!(queue.pop(), Some(1));
    }
}
